### ISSUE TITLE
"Only Once" example hangs

### ISSUE THREAD

**Issue by armanbilge** [typelevel/cats-effect#3283]
Example from https://typelevel.org/cats-effect/docs/std/deferred#only-once

This is a regression since 3.3.14. It seems to run the first time, and then hang after for all subsequent invocations.

```scala
//> using lib "org.typelevel::cats-core::2.9.0"
//> using lib "org.typelevel::cats-effect::3.4.1"

import cats.effect._
import cats.syntax.all._

object App extends IOApp.Simple {
  def start(d: Deferred[IO, Int]): IO[Unit] = {
    val attemptCompletion: Int => IO[Unit] = n => d.complete(n).void

    List(
      IO.race(attemptCompletion(1), attemptCompletion(2)),
      d.get.flatMap { n => IO(println(show"Result: $n")) }
    ).parSequence.void
  }

  def run: IO[Unit] =
    for {
      d <- Deferred[IO, Int]
      _ <- start(d)
    } yield ()
}
```

**Comment by armanbilge** [comment]
Seems to be specific to Scala 3.

**Comment by mbesida** [comment]
@armanbilge some pecularities how scala 3 is compiled to bytecode? Because even compiled jar started with `java -jar` reproduces this

**Comment by armanbilge** [comment]
I just reproduced with 3.3.12 🤔  edit: and 3.3.14. So maybe it's not a regression.

**Comment by armanbilge** [comment]
So after more investigation, I think I've traced this as far back as CE 3.2.1. This must be some sort of a race condition since it's tricky to reproduce: I tested each version by executing the program on loop until it hangs, often taking several minutes. So far I have not seen it hang with CE 3.2.0. I have also never seen it hang with Scala 2.13.

This is the exact code I last reproduced with.
<details>

```scala
//> using scala "3.2.1"
//> using lib "org.typelevel::cats-core::2.9.0"
//> using lib "org.typelevel::cats-effect::3.2.1"

import cats.effect._
import cats.syntax.all._

object App extends IOApp.Simple {
  def start(d: Deferred[IO, Int]): IO[Unit] = {
    val attemptCompletion: Int => IO[Unit] = n => d.complete(n).void

    List(
      IO.race(attemptCompletion(1), attemptCompletion(2)),
      d.get.flatMap { n => IO.println(show"Result: $n") }
    ).parSequence.void
  }

  def run: IO[Unit] =
    for {
      d <- Deferred[IO, Int]
      _ <- start(d)
    } yield ()
}
```

</details>

Looking at the changelog for CE 3.2.1, it _could_ be related to this PR maybe?
- https://github.com/typelevel/cats-effect/pull/2159

**Comment by armanbilge** [comment]
One more datapoint: assuming it is Scala 3 specific, the issue is not with the published artifacts, but something about the code itself. It still hangs even when using the 2.13 Cats Effect artifacts with the Scala 3 compiler.

```
//> using scala "3.2.1"
//> using lib "org.typelevel:cats-core_2.13:2.9.0"
//> using lib "org.typelevel:cats-effect_2.13:3.2.1"
```

**Comment by durban** [comment]
This is interesting :-)

The important part is this:
```scala
IO.race(attemptCompletion(1), attemptCompletion(2))
```

I think what happens is that:

1. Let's say `attemptCompletion(1)` is faster, changes the contents of the `AtomicReference`.
2. `attemptCompletion(2)` runs on another thread, sees the completed state of the `Deferred`, so it succeeds with a `false` result, and _wins the `race`_.
3. `attemptCompletion(1)` is cancelled (because it lost the race) before it could wake up any readers of the `Deferred`.
4. There is only one reader: the `get`. Thus, it deadlocks.

I didn't try it, but possibly `Deferred#complete` should be uncancellable?

(Btw, I had no trouble reproducing this with scala 2.13...)

**Comment by djspiewak** [comment]
@durban Wow… Stunning work. `complete` does indeed need to be uncancelable. You wanna PR so you get the credit? Otherwise I have it up right now. Mad kudos on the diagnosis.

**Comment by durban** [comment]
@djspiewak Thank you for the kind words. I've created #3290.

**Comment by SystemFw** [comment]
I've generally defaulted to making the whole "write path" of things (which tend to complete deferreds) uncancelable, but given that the cost of nested uncancelables is low, having `complete` be uncancelable sounds like a better default
